"My first dapp, check readme"

PrincesoDan opened the pull request.

Commit:
"My first dapp, check readme"

esteblock commented:
Preview: https://create-soroban-dapp-git-fork-princesodan-princesodan-paltalabs.vercel.app/

Commit:
improve_challenge_title

PrincesoDan commented:
I stayed motivated by the challenge, so I made some additional improvements to my initial code c:
https://github.com/PrincesoDan/paltaLabsChallenge/tree/PrincesoDan

@esteblock

esteblock commented on soroban-react-dapp/.soroban/identity/admin.toml:
Very bad practice. Might lead to loss of funds or hack

esteblock commented on soroban-react-dapp/README.md:
Bad practice to hard code publicly secret keys

esteblock commented on soroban-react-dapp/contracts/deployments.json:
```diff
@@ -7,6 +7,11 @@
+  {
+    "contractId": "title",
+    "networkPassphrase": "Test SDF Network ; September 2015",
+    "contractAddress": "CAS3V3WZFZYQMHFXRJJOCYT2JMLO6CFZSIQK64NN4WR3WN4OB5TXV3XL"
```
https://stellar.expert/explorer/testnet/contract/CAS3V3WZFZYQMHFXRJJOCYT2JMLO6CFZSIQK64NN4WR3WN4OB5TXV3XL

esteblock commented on soroban-react-dapp/contracts/package.json:
```diff
@@ -5,7 +5,8 @@
     "build": "tsc",
-    "deploy": "yarn build && node dist/scripts/deploy.js"
+    "deploy": "yarn build && node dist/scripts/deploy.js",
+    "testtitle": "yarn build && node dist/scripts/test_title.js"
```
deploy:testitle?

PrincesoDan replied:
yes

esteblock commented on soroban-react-dapp/contracts/title/src/lib.rs:
```diff
@@ -0,0 +1,85 @@
+
+    // Adds an authorized address to modify the title
+    pub fn add_user(env: Env, user: Address) {
+        let admin: Address = env.storage().instance().get(&ADMIN).unwrap();
```
this will panic if admin is not defined

esteblock commented on soroban-react-dapp/contracts/title/src/lib.rs:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    // Adds an authorized address to modify the title
+    pub fn add_user(env: Env, user: Address) {
```
Because you are using a mapping user>bool you could put true or false and remove a user direclty here

esteblock commented on soroban-react-dapp/contracts/title/src/lib.rs:
You are never making reference to test
Indeed cargo test will run no test

![image](https://github.com/user-attachments/assets/db88975d-7554-4a23-bcb5-e1261190ccf2)

esteblock replied:
If I add `mod test;` at the bottom of `lib.rs` I get bunch of errors

```rust
root@83d6e5f3920c:/workspace/contracts/title# cargo test
   Compiling title v0.0.1 (/workspace/contracts/title)
error[E0603]: struct `Vec` is private
  --> src/test.rs:4:60
   |
4  | use soroban_sdk::{Env, testutils::{Address as TestAddress, Vec as TestVec}, Symbol, String};
   |                                                            ^^^ private struct
   |
note: the struct `Vec` is defined here
  --> /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/soroban-sdk-21.1.1/src/testutils.rs:18:28
   |
18 | use crate::{xdr, Env, Val, Vec};
   |                            ^^^
help: import `Vec` directly
   |
4  | use soroban_sdk::{Env, testutils::{Address as TestAddress, soroban_sdk::vec::Vec}, Symbol, String};
   |                                                            ~~~~~~~~~~~~~~~~~~~~~

warning: unused import: `Symbol`
 --> src/test.rs:4:77
  |
4 | use soroban_sdk::{Env, testutils::{Address as TestAddress, Vec as TestVec}, Symbol, String};
  |                                                                             ^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0782]: trait objects must include the `dyn` keyword
 --> src/test.rs:9:17
  |
9 |     let admin = TestAddress::random(&env);
  |                 ^^^^^^^^^^^
  |
help: add `dyn` keyword before this trait
  |
9 |     let admin = <dyn TestAddress>::random(&env);
  |                 ++++            +

error[E0599]: no function or associated item named `random` found for trait `soroban_sdk::testutils::Address`
 --> src/test.rs:9:30
  |
9 |     let admin = TestAddress::random(&env);
  |                              ^^^^^^ function or associated item not found in `Address`

error[E0782]: trait objects must include the `dyn` keyword
  --> src/test.rs:24:17
   |
24 |     let admin = TestAddress::random(&env);
   |                 ^^^^^^^^^^^
   |
help: add `dyn` keyword before this trait
   |
24 |     let admin = <dyn TestAddress>::random(&env);
   |                 ++++            +

error[E0599]: no function or associated item named `random` found for trait `soroban_sdk::testutils::Address`
  --> src/test.rs:24:30
   |
24 |     let admin = TestAddress::random(&env);
   |                              ^^^^^^ function or associated item not found in `Address`

error[E0782]: trait objects must include the `dyn` keyword
  --> src/test.rs:25:16
   |
25 |     let user = TestAddress::random(&env);
   |                ^^^^^^^^^^^
   |
help: add `dyn` keyword before this trait
   |
25 |     let user = <dyn TestAddress>::random(&env);
   |                ++++            +

error[E0599]: no function or associated item named `random` found for trait `soroban_sdk::testutils::Address`
  --> src/test.rs:25:29
   |
25 |     let user = TestAddress::random(&env);
   |                             ^^^^^^ function or associated item not found in `Address`

error[E0782]: trait objects must include the `dyn` keyword
  --> src/test.rs:38:17
   |
38 |     let admin = TestAddress::random(&env);
   |                 ^^^^^^^^^^^
   |
help: add `dyn` keyword before this trait
   |
38 |     let admin = <dyn TestAddress>::random(&env);
   |                 ++++            +

error[E0599]: no function or associated item named `random` found for trait `soroban_sdk::testutils::Address`
  --> src/test.rs:38:30
   |
38 |     let admin = TestAddress::random(&env);
   |                              ^^^^^^ function or associated item not found in `Address`

error[E0782]: trait objects must include the `dyn` keyword
  --> src/test.rs:39:16
   |
39 |     let user = TestAddress::random(&env);
   |                ^^^^^^^^^^^
   |
help: add `dyn` keyword before this trait
   |
39 |     let user = <dyn TestAddress>::random(&env);
   |                ++++            +

error[E0599]: no function or associated item named `random` found for trait `soroban_sdk::testutils::Address`
  --> src/test.rs:39:29
   |
39 |     let user = TestAddress::random(&env);
   |                             ^^^^^^ function or associated item not found in `Address`

error[E0782]: trait objects must include the `dyn` keyword
  --> src/test.rs:55:17
   |
55 |     let admin = TestAddress::random(&env);
   |                 ^^^^^^^^^^^
   |
help: add `dyn` keyword before this trait
   |
55 |     let admin = <dyn TestAddress>::random(&env);
   |                 ++++            +

error[E0599]: no function or associated item named `random` found for trait `soroban_sdk::testutils::Address`
  --> src/test.rs:55:30
   |
55 |     let admin = TestAddress::random(&env);
   |                              ^^^^^^ function or associated item not found in `Address`

error[E0782]: trait objects must include the `dyn` keyword
  --> src/test.rs:56:21
   |
56 |     let new_admin = TestAddress::random(&env);
   |                     ^^^^^^^^^^^
   |
help: add `dyn` keyword before this trait
   |
56 |     let new_admin = <dyn TestAddress>::random(&env);
   |                     ++++            +

error[E0599]: no function or associated item named `random` found for trait `soroban_sdk::testutils::Address`
  --> src/test.rs:56:34
   |
56 |     let new_admin = TestAddress::random(&env);
   |                                  ^^^^^^ function or associated item not found in `Address`

```

esteblock commented on soroban-react-dapp/contracts/title/src/lib.rs:
```diff
@@ -0,0 +1,85 @@
+    // Retrieves the contract's administrator address
+    pub fn address_admin(env: Env) -> Result<Address, ContractError> {
+        match env.storage().instance().get(&ADMIN) {
+            Some(admin_address) => Ok(admin_address),
```
I am not sure, because I did not test it, but is my guess that this should be 

`Some(admin_address) => admin_address,`

Because the compiler does not know what the storage instance is saving... you are not getting any warning at moment of compiling

esteblock commented on soroban-react-dapp/contracts/title/src/test.rs:
```diff
@@ -0,0 +1,63 @@
+}
+
+#[test]
+fn test_modify_title() {
```
There is no test to indeed see that the auth control is working

esteblock commented on soroban-react-dapp/README.md:
```diff
@@ -1,3 +1,59 @@
+
+bash
+
+yarn deploy testnet title
```
cd contracts

esteblock replied:
No contracts build instructions

![image](https://github.com/user-attachments/assets/df1fdf3f-2c2c-4469-a603-b87c0f71db14)

Commit:
fix title.rs and test.rs

Commit:
feat: Update README with deployment and admin setup instructions, improve ManageTitle component

- Updated README.md with comprehensive instructions on setting up the application, deploying the smart contract, and configuring the admin user.
- Deployed the smart contract to the Stellar blockchain using Soroban.
- Enhanced ManageTitle.tsx component:
  - Implemented automatic refresh for the list of authorized users.
  - Added automatic form field clearing after actions like adding a user, modifying the admin, or changing the title.
  - Improved semantic HTML structure: set contract section titles with larger, bold font.

esteblock commented on soroban-react-dapp/contracts/title/src/test.rs:
```diff
@@ -0,0 +1,167 @@
+    let user = Address::generate(&env);
+
+    // Autenticar y setear al admin
+    client.mock_auths(&[MockAuth {
```
Esta bien mockeado el auth, pero para testear se deberían revisar las ultimas auths y confirmar que la auth realmente existió. Esto esta mejor que antes de toda maneras

PrincesoDan committed:
Update README.md
